Sources/apollo-ios-pagination/GraphQLQueryPager.swift

```diff
@@ -0,0 +1,212 @@
+  public enum PageExtractionData {
+    case initial(InitialQuery.Data)
+    case paginated(PaginatedQuery.Data)
+  }
+
+  /// Whether or not we can load more information based on the current page.
+  public var canLoadNext: Bool { currentPageInfo?.canLoadMore ?? false }
+
+  public typealias Output = (InitialQuery.Data, [PaginatedQuery.Data], UpdateSource)
+
+  private let client: any ApolloClientProtocol
+  private var firstPageWatcher: GraphQLQueryWatcher<InitialQuery>?
+  private var nextPageWatchers: [GraphQLQueryWatcher<PaginatedQuery>] = []
+  private let initialQuery: InitialQuery
+  private let nextPageResolver: (PaginationInfo) -> PaginatedQuery?
+  private let extractPageInfo: (PageExtractionData) -> PaginationInfo
+  private var currentPageInfo: PaginationInfo? {
+    guard let last = pageOrder.last else { return nil }
+    if let data = varMap[last] {
+      return extractPageInfo(.paginated(data))
+    } else if let initialPageResult {
+      return extractPageInfo(.initial(initialPageResult))
+    } else {
+      return nil
+    }
+  }
+
+  private var onUpdate: ((Output) -> Void)?
+  private var onError: ((Error) -> Void)?
+
+  private var initialPageResult: InitialQuery.Data?
+  private var latest: (InitialQuery.Data, [PaginatedQuery.Data])? {
+    guard let initialPageResult else { return nil }
+    return (initialPageResult, pageOrder.compactMap({ varMap[$0] }))
+  }
+
+  /// Array of page info used to fetch next pages. Maintains an order of values used to fetch each page in a connection.
+  private var pageOrder = [AnyHashable]()
+
+  /// Maps each query variable set to latest results from internal watchers.
+  private var varMap: [AnyHashable: PaginatedQuery.Data] = [:]
+
+  private var activeTask: Task<Void, Never>?
+
+  /// Designated Initializer
+  /// - Parameters:
+  ///   - client: Apollo Client
+  ///   - initialQuery: The initial query that is being watched
+  ///   - extractPageInfo: The `PageInfo` derived from `PageExtractionData`
+  ///   - nextPageResolver: The resolver that can derive the query for loading more. This can be a different query than the `initialQuery`.
+  ///   - onError: The callback when there is an error.
+  public init<P: PaginationInfo>(
+    client: ApolloClientProtocol,
+    initialQuery: InitialQuery,
+    extractPageInfo: @escaping (PageExtractionData) -> P,
+    nextPageResolver: @escaping (P) -> PaginatedQuery
+  ) {
+    self.client = client
+    self.initialQuery = initialQuery
+    self.extractPageInfo = extractPageInfo
+    self.nextPageResolver = { page in
+      guard let page = page as? P else { return nil }
+      return nextPageResolver(page)
+    }
+  }
+
+  // MARK: - Public API
+
+  /// Subscribe to new data from this watcher.
+  /// - Returns: An async stream that can be iterated over.
+  public func subscribe() -> AsyncStream<Result<Output, Error>> {
+    AsyncStream { continuation in
+      self.onUpdate = { continuation.yield(.success($0)) }
```

Comment:
Thinking further, `subscribe()` won't retain the existing pages, only send new ones since you subscribed. 

I'm going back on forth on what the API should look like here. Maybe we should try to provide a few different ones?

When `InitialQuery` and `PaginatedQuery` are different, then we need to keep `Output` as it's defined `(InitialQuery.Data, [PaginatedQuery.Data], UpdateSource)`. But if `InitialQuery == PaginatedQuery`, then we can really have an `Output` that is just `[PaginatedQuery.Data]`. In that case, we could create an extension that provides different functions.

1. A stream of all existing pages that have already been fetched should be sent immediately, and then new pages as they are fetched. Basically this is a functional reactive replay stream, which doesn't exist natively in `Combine`, but could be easily created.

2. Just a subscription to new pages that are fetched, not sending all the already fetched pages. I think this would be a returning an `AnyPublisher` that wraps an underlying `PassthroughSubject`.

These don't need to be created as part of the MVP, but it's worth noting that it's something we can add!

Reply:
> If you call subscribe twice, the first subscriber will no longer get updates or be terminated in this situation. I'm thinking we should use Combine here to create a single PassthroughSubject that can be subscribed to by multiple subscribers instead.

It's a great callout -- using a passthrough subject seems like the right move.